Fix OpenGOAL aarch64 Linux build.

The upstream tree has partial ARM64 support, but the Linux/aarch64 build still trips over x86-only debug register access, x86-only mmap flags, Apple-flavored ARM assembly syntax, and OpenSSL static library selection in Buildroot sysroots.

Upstream: not submitted

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -53,7 +53,7 @@
         -fcxx-exceptions \
         -fexceptions \
         -fdiagnostics-color=always \
-        -mavx \
+        \
         -Wall \
         -Wno-c++11-narrowing \
         -Wno-c++98-compat \
@@ -108,7 +108,7 @@
         -Wshadow \
         -Wsign-promo \
         -fdiagnostics-color=always \
-        -mavx")
+       ")
 
     # additional c++ flags for release mode for our projects
     if(CMAKE_BUILD_TYPE MATCHES "Release")
@@ -137,7 +137,7 @@
 
     # TODO - make a proper flag for arm compiling
     if (CMAKE_APPLE_SILICON_PROCESSOR STREQUAL "x86_64")
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mavx")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
     else()
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=native -mcrc")
     endif()
@@ -218,9 +218,14 @@
 elseif(APPLE)
     set(CURL_USE_SECTRANSP ON) # native macOS SSL support
 else()
-    if(STATICALLY_LINK)
+    if(STATICALLY_LINK AND NOT CMAKE_SYSTEM_PROCESSOR MATCHES "^(arm64|aarch64)$")
         set(OPENSSL_USE_STATIC_LIBS TRUE)
     endif()
+    if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(arm64|aarch64)$" AND CMAKE_SYSROOT)
+        set(OPENSSL_USE_STATIC_LIBS FALSE)
+        set(OPENSSL_SSL_LIBRARY "${CMAKE_SYSROOT}/usr/lib/libssl.so" CACHE FILEPATH "OpenSSL SSL library" FORCE)
+        set(OPENSSL_CRYPTO_LIBRARY "${CMAKE_SYSROOT}/usr/lib/libcrypto.so" CACHE FILEPATH "OpenSSL crypto library" FORCE)
+    endif()
     set(CURL_USE_OPENSSL ON) # not native, but seems to be the best choice for linux
 endif()
 include_directories(third-party/curl/include)
--- a/common/cross_os_debug/xdbg.cpp
+++ b/common/cross_os_debug/xdbg.cpp
@@ -24,6 +24,11 @@
 #include <sys/types.h>
 #include <sys/user.h>
 #include <sys/wait.h>
+#if defined(__aarch64__)
+#include <asm/ptrace.h>
+#include <elf.h>
+#include <sys/uio.h>
+#endif
 #elif _WIN32
 #define NOMINMAX
 #define WIN32_LEAN_AND_MEAN
@@ -211,6 +216,20 @@
  * Get all registers now. Must be attached and stopped
  */
 bool get_regs_now(const ThreadID& tid, Regs* out) {
+#if defined(__aarch64__)
+  user_pt_regs regs = {};
+  iovec iov = {&regs, sizeof(regs)};
+  if (ptrace(PTRACE_GETREGSET, tid.id, NT_PRSTATUS, &iov) < 0) {
+    printf("[Debugger] Failed to PTRACE_GETREGSET %s\n", strerror(errno));
+    return false;
+  }
+
+  for (int i = 0; i < 16; i++) {
+    out->gprs[i] = regs.regs[i];
+  }
+  out->rip = regs.pc;
+  return true;
+#else
   user regs = {};
   if (ptrace(PTRACE_GETREGS, tid.id, nullptr, &regs) < 0) {
     printf("[Debugger] Failed to PTRACE_GETREGS %s\n", strerror(errno));
@@ -237,12 +256,32 @@
 
   // todo, get fprs.
   return true;
+#endif
 }
 
 /*!
  * Set all registers now. Must be attached and stopped
  */
 bool set_regs_now(const ThreadID& tid, const Regs& out) {
+#if defined(__aarch64__)
+  user_pt_regs regs = {};
+  iovec iov = {&regs, sizeof(regs)};
+  if (ptrace(PTRACE_GETREGSET, tid.id, NT_PRSTATUS, &iov) < 0) {
+    printf("[Debugger] Failed to PTRACE_GETREGSET %s\n", strerror(errno));
+    return false;
+  }
+
+  for (int i = 0; i < 16; i++) {
+    regs.regs[i] = out.gprs[i];
+  }
+  regs.pc = out.rip;
+
+  if (ptrace(PTRACE_SETREGSET, tid.id, NT_PRSTATUS, &iov) < 0) {
+    printf("[Debugger] Failed to PTRACE_SETREGSET %s\n", strerror(errno));
+    return false;
+  }
+  return true;
+#else
   user regs = {};
   if (ptrace(PTRACE_GETREGS, tid.id, nullptr, &regs) < 0) {
     printf("[Debugger] Failed to PTRACE_GETREGS %s\n", strerror(errno));
@@ -273,6 +312,7 @@
   }
   // todo, set fprs.
   return true;
+#endif
 }
 
 /*!
--- a/game/CMakeLists.txt
+++ b/game/CMakeLists.txt
@@ -1,5 +1,5 @@
 # Set a more convenient ARM flag
-if(CMAKE_SYSTEM_PROCESSOR MATCHES "arm64")
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(arm64|aarch64)$")
     set(ARM64_ARCH TRUE)
     message(STATUS "ARM64 architecture detected")
 else()
@@ -12,7 +12,7 @@
     set(OG_ASM_FUNCS_FILE kernel/asm_funcs_arm64.s)
     enable_language(ASM)
     set(CMAKE_ASM_SOURCE_FILE_EXTENSIONS ${CMAKE_ASM_SOURCE_FILE_EXTENSIONS} s)
-    set_source_files_properties(${OG_ASM_FUNCS_FILE} PROPERTIES COMPILE_FLAGS "-arch arm64 -g")
+    set_source_files_properties(${OG_ASM_FUNCS_FILE} PROPERTIES COMPILE_FLAGS "-g")
 else()
     set(OG_ASM_FUNCS_FILE kernel/asm_funcs_x86_64.asm)
     enable_language(ASM_NASM)
--- a/game/kernel/asm_funcs_arm64.s
+++ b/game/kernel/asm_funcs_arm64.s
@@ -1,16 +1,16 @@
-;; GOAL Runtime assembly functions. These exist only in the arm64 version of GOAL.
-;; - https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms#Pass-arguments-to-functions-correctly
-;; - https://en.wikipedia.org/wiki/Calling_convention#ARM_(A64)
-;; - https://student.cs.uwaterloo.ca/~cs452/docs/rpi4b/aapcs64.pdf
-;; - s16–s31 (d8–d15, q4–q7) must be preserved
-;; - s0–s15 (d0–d7, q0–q3) and d16–d31 (q8–q15) do not need to be preserved
-;; - https://devblogs.microsoft.com/oldnewthing/20220728-00/?p=106912
-;; - ;; - https://courses.cs.washington.edu/courses/cse469/19wi/arm64.pdf
+//// GOAL Runtime assembly functions. These exist only in the arm64 version of GOAL.
+//// - https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms#Pass-arguments-to-functions-correctly
+//// - https://en.wikipedia.org/wiki/Calling_convention#ARM_(A64)
+//// - https://student.cs.uwaterloo.ca/~cs452/docs/rpi4b/aapcs64.pdf
+//// - s16–s31 (d8–d15, q4–q7) must be preserved
+//// - s0–s15 (d0–d7, q0–q3) and d16–d31 (q8–q15) do not need to be preserved
+//// - https://devblogs.microsoft.com/oldnewthing/20220728-00/?p=106912
+//// - //// - https://courses.cs.washington.edu/courses/cse469/19wi/arm64.pdf
 
 .text
 
-;; Call C++ code on arm64 systems, from GOAL.
-;; Following the macOS documentation which mostly aligns with standard arm64
+//// Call C++ code on arm64 systems, from GOAL.
+//// Following the macOS documentation which mostly aligns with standard arm64
 .global _arg_call_arm64
 .align 4
 _arg_call_arm64:
@@ -18,8 +18,8 @@
   mov	x29, sp
   ldr x8, [sp], #16
 
-  ; Putting an exclamation point after the close-bracket 
-  ; means that the calculated effective address is written back to the base register. (pre-indexing)
+  // Putting an exclamation point after the close-bracket 
+  // means that the calculated effective address is written back to the base register. (pre-indexing)
   stp q15, q14, [sp, #-32]!
   stp q13, q12, [sp, #-32]!
   stp q11, q10, [sp, #-32]!
@@ -36,12 +36,12 @@
   ret
 
 
-;; Call C++ code on arm64 systems, from GOAL. 
-;; 
-;; Put arguments on the stack and put a pointer to this array in the first arg.
-;; this function pushes all 8 OpenGOAL registers into a stack array.
-;; then it calls the function pointed to by x0 (RAX in x86) with a pointer to this array.
-;; it returns the return value of the called function.
+//// Call C++ code on arm64 systems, from GOAL. 
+//// 
+//// Put arguments on the stack and put a pointer to this array in the first arg.
+//// this function pushes all 8 OpenGOAL registers into a stack array.
+//// then it calls the function pointed to by x0 (RAX in x86) with a pointer to this array.
+//// it returns the return value of the called function.
 .global _stack_call_arm64
 .align 4
 _stack_call_arm64:
@@ -54,25 +54,25 @@
   stp q11, q10, [sp, #-32]!
   stp q9, q8, [sp, #-32]!
 
-  ; create stack array of arguments
-  ; arg 7 (R11 in x86)
-  ; arg 6 (R10 in x86)
-  ; arg 5 (R8 in x86)
-  ; arg 4 (R8 in x86)
-  ; arg 3 (RCX in x86)
-  ; arg 2 (RDX in x86)
-  ; arg 1 (RSI in x86)
-  ; arg 0 (RDI in x86)
+  // create stack array of arguments
+  // arg 7 (R11 in x86)
+  // arg 6 (R10 in x86)
+  // arg 5 (R8 in x86)
+  // arg 4 (R8 in x86)
+  // arg 3 (RCX in x86)
+  // arg 2 (RDX in x86)
+  // arg 1 (RSI in x86)
+  // arg 0 (RDI in x86)
   stp x7, x6, [sp, #-16]!
   stp x5, x4, [sp, #-16]!
   stp x3, x2, [sp, #-16]!
   stp x1, x0, [sp, #-16]!
 
-  ; set first argument
+  // set first argument
   mov x19, sp
-  ; call function
+  // call function
   blr x8
-  ; restore arguments
+  // restore arguments
   ldp x1, x0, [sp], #16
   ldp x3, x2, [sp], #16
   ldp x5, x4, [sp], #16
@@ -84,116 +84,116 @@
   ldp q14, q15, [sp], #32
 
   ldp	x29, x30, [sp], #16
-  ; return!
+  // return!
   ret
 
-;; Call c++ code through mips2c.
-;; GOAL will call a dynamically generated trampoline.
-;; The trampoline will have pushed the exec function and stack offset onto the stack
+//// Call c++ code through mips2c.
+//// GOAL will call a dynamically generated trampoline.
+//// The trampoline will have pushed the exec function and stack offset onto the stack
 .global _mips2c_call_arm64
 .align 4
 _mips2c_call_arm64:
   stp	x29, x30, [sp, #-16]!
   mov	x29, sp
-  ;; TODO - this is really weird using half an XMM, this makes the arm assembly
-  ;; more difficult - this probably isn't required for arm?
-  ;; grab the address to call and put it in xmm0
-  ;; TODO - this stack pointer manipulation might be a problem for ARM64 which requires 16byte alignment
-  ;; sub sp, 8
+  //// TODO - this is really weird using half an XMM, this makes the arm assembly
+  //// more difficult - this probably isn't required for arm?
+  //// grab the address to call and put it in xmm0
+  //// TODO - this stack pointer manipulation might be a problem for ARM64 which requires 16byte alignment
+  //// sub sp, 8
   ldr q0, [sp, #+16]
-  ;; grab the stack offset
+  //// grab the stack offset
   ldr x0, [sp, #+8]
 
-  ;; first, save quadword registers
+  //// first, save quadword registers
   stp q15, q14, [sp, #-32]!
   stp q13, q12, [sp, #-32]!
   stp q11, q10, [sp, #-32]!
   stp q9, q8, [sp, #-32]!
 
-  ; NOTE - in x86 the 2 special registers are saved (R10 and R11)
-  ; we don't need to do that in ARM64, there are plenty of registers to work with
+  // NOTE - in x86 the 2 special registers are saved (R10 and R11)
+  // we don't need to do that in ARM64, there are plenty of registers to work with
 
-  ;; oof
+  //// oof
   sub sp, sp, 1280
-  str x0, [sp, #+64] ; arg 0 (RDI in x86) and 
-  str x1, [sp, #+80] ; arg 1 (RSI in x86)
-  str x2, [sp, #+96] ; arg 2 (RDX in x86) and arg 3 (RCX in x86)
-  str x3, [sp, #+112] ; arg 2 (RDX in x86) and arg 3 (RCX in x86)
-  str x4, [sp, #+128] ; arg 4 (R8 in x86) and arg 5 (R8 in x86)
-  str x5, [sp, #+144] ; arg 4 (R8 in x86) and arg 5 (R8 in x86)
-  str x6, [sp, #+160] ; arg 6 (R10 in x86) and arg 7 (R11 in x86)
-  str x7, [sp, #+176] ; arg 6 (R10 in x86) and arg 7 (R11 in x86)
-  str x20, [sp, #+352] ;; s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
-  str x21, [sp, #+368] ;; s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
-
-  mov x0, sp ; move the stack pointer to arg 0
-  sub x0, x0, x22 ; R15 is a "special" offset TODO - whats special about it?
-  str x0, [sp, #+464] ;; mip2c code's MIPS stack
-
-  mov x0, sp ;; move the stack pointer to the new position
-
-  sub sp, sp, x8 ;; allocate space on the stack for GOAL fake stack
-  stp x8, x8, [sp, #-16]! ;; and remember this so we can find our way back
-
-  ;; TODO - this used to be a movq rax, xmm0
-  ;; TODO - not sure why an `xmm` was used because that movq only uses the lower 64bits anyway
-  mov x0, v0.d[0] ; represents the lower 64 bits of q0
-  blr x8 ;; call!
+  str x0, [sp, #+64] // arg 0 (RDI in x86) and 
+  str x1, [sp, #+80] // arg 1 (RSI in x86)
+  str x2, [sp, #+96] // arg 2 (RDX in x86) and arg 3 (RCX in x86)
+  str x3, [sp, #+112] // arg 2 (RDX in x86) and arg 3 (RCX in x86)
+  str x4, [sp, #+128] // arg 4 (R8 in x86) and arg 5 (R8 in x86)
+  str x5, [sp, #+144] // arg 4 (R8 in x86) and arg 5 (R8 in x86)
+  str x6, [sp, #+160] // arg 6 (R10 in x86) and arg 7 (R11 in x86)
+  str x7, [sp, #+176] // arg 6 (R10 in x86) and arg 7 (R11 in x86)
+  str x20, [sp, #+352] //// s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
+  str x21, [sp, #+368] //// s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
+
+  mov x0, sp // move the stack pointer to arg 0
+  sub x0, x0, x22 // R15 is a "special" offset TODO - whats special about it?
+  str x0, [sp, #+464] //// mip2c code's MIPS stack
+
+  mov x0, sp //// move the stack pointer to the new position
+
+  sub sp, sp, x8 //// allocate space on the stack for GOAL fake stack
+  stp x8, x8, [sp, #-16]! //// and remember this so we can find our way back
+
+  //// TODO - this used to be a movq rax, xmm0
+  //// TODO - not sure why an `xmm` was used because that movq only uses the lower 64bits anyway
+  mov x0, v0.d[0] // represents the lower 64 bits of q0
+  blr x8 //// call!
 
-  ;; unallocate
+  //// unallocate
   ldp x8, x8, [sp], #16
   add sp, sp, x8
 
   ldr x8, [sp, #+32]
 
-  add sp, sp, 1280 ; reset the stackpointer back
+  add sp, sp, 1280 // reset the stackpointer back
 
   ldp q9, q8, [sp], #32
   ldp q10, q11, [sp], #32
   ldp q12, q13, [sp], #32
   ldp q14, q15, [sp], #32
 
-  add sp, sp, 24 ;; 16 for the stuff pushed by trampoline
+  add sp, sp, 24 //// 16 for the stuff pushed by trampoline
   ldp	x29, x30, [sp], #16
   ret
 
-;; The _call_goal_asm function is used to call a GOAL function from C.
-;; It calls on the parent stack, which is a bad idea if your stack is not already a GOAL stack.
-;; It supports up to 3 arguments and a return value.
-;; This should be called with the arguments:
-;; - first goal arg
-;; - second goal arg
-;; - third goal arg
-;; - address of function to call
-;; - address of the symbol table
-;; - GOAL memory space offset
+//// The _call_goal_asm function is used to call a GOAL function from C.
+//// It calls on the parent stack, which is a bad idea if your stack is not already a GOAL stack.
+//// It supports up to 3 arguments and a return value.
+//// This should be called with the arguments:
+//// - first goal arg
+//// - second goal arg
+//// - third goal arg
+//// - address of function to call
+//// - address of the symbol table
+//// - GOAL memory space offset
 .global _call_goal_asm_arm64
 .align 4
 _call_goal_asm_arm64:
   stp	x29, x30, [sp, #-16]!
   mov	x29, sp
-  ;; saved registers we need to modify for GOAL should be preserved
-  ; ARM64 requires 16-byte stack pointer alignment
+  //// saved registers we need to modify for GOAL should be preserved
+  // ARM64 requires 16-byte stack pointer alignment
   stp x20, x21, [sp, #-16]!
   str x22, [sp, #-16]!
 
-  ;; x0 - first arg
-  ;; x1 - second arg
-  ;; x2 - third arg
-  ;; x3 - function pointer
-  ;; x4 - st (goes in x20 and x21)
-  ;; x5 - off (goes in x22)
+  //// x0 - first arg
+  //// x1 - second arg
+  //// x2 - third arg
+  //// x3 - function pointer
+  //// x4 - st (goes in x20 and x21)
+  //// x5 - off (goes in x22)
 
-  ;; set GOAL process
+  //// set GOAL process
   mov x20, x4
-  ;; symbol table
+  //// symbol table
   mov x21, x4
-  ;; offset
+  //// offset
   mov x22, x5
-  ;; call GOAL by function pointer
+  //// call GOAL by function pointer
   blr x3
 
-  ;; restore saved registers.
+  //// restore saved registers.
   ldr x22, [sp], #16
   ldp x20, x21, [sp], #16
   ldp	x29, x30, [sp], #16
@@ -204,75 +204,75 @@
 _call_goal8_asm_arm64:
   stp	x29, x30, [sp, #-16]!
   mov	x29, sp
-  ;; saved registers we need to modify for GOAL should be preserved
-  ; ARM64 requires 16-byte stack pointer alignment
+  //// saved registers we need to modify for GOAL should be preserved
+  // ARM64 requires 16-byte stack pointer alignment
   stp x20, x21, [sp, #-16]!
   str x22, [sp, #-16]!
 
-  ;; x0 - first arg (func)
-  ;; x1 - second arg (arg array)
-  ;; x2 - third arg  (0)
-  ;; x3 - pp (goes in r13)
-  ;; x4  - st (goes in r14)
-  ;; x5  - off (goes in r15)
+  //// x0 - first arg (func)
+  //// x1 - second arg (arg array)
+  //// x2 - third arg  (0)
+  //// x3 - pp (goes in r13)
+  //// x4  - st (goes in r14)
+  //// x5  - off (goes in r15)
 
-  ;; set GOAL function pointer
+  //// set GOAL function pointer
   mov x20, x3
-  ;; st
+  //// st
   mov x21, x4
-  ;; offset
+  //// offset
   mov x22, x5
-  ;; move function to temp
+  //// move function to temp
   mov x8, x0
-  ;; extract arguments
-  ldr x0, [x1]  ;; 0
-  ldr x2, [x1, #+16] ;; 2
-  ldr x3, [x1, #+24] ;; 3
-  ldr x4, [x1, #+32]  ;; 4
-  ldr x5, [x1, #+40]  ;; 5
-  ldr x6, [x1, #+48] ;; 6
-  ldr x7, [x1, #+56]  ;; 7
-  ldr x1, [x1, #+8] ;; 1 (do this last)
-  ;; call GOAL by function pointer
+  //// extract arguments
+  ldr x0, [x1]  //// 0
+  ldr x2, [x1, #+16] //// 2
+  ldr x3, [x1, #+24] //// 3
+  ldr x4, [x1, #+32]  //// 4
+  ldr x5, [x1, #+40]  //// 5
+  ldr x6, [x1, #+48] //// 6
+  ldr x7, [x1, #+56]  //// 7
+  ldr x1, [x1, #+8] //// 1 (do this last)
+  //// call GOAL by function pointer
   blr x8
 
-  ;; retore registers.
+  //// retore registers.
   ldr x22, [sp], #16
   ldp x20, x21, [sp], #16
   ldp	x29, x30, [sp], #16
   ret
 
-;; Call goal, but switch stacks.
+//// Call goal, but switch stacks.
 .global _call_goal_on_stack_asm_arm64
 .align 4
 _call_goal_on_stack_asm_arm64:
   stp	x29, x30, [sp, #-16]!
   mov	x29, sp
-  ;; x0 - stack pointer
-  ;; x1 - unused
-  ;; x2 - unused
-  ;; x3 - function pointer
-  ;; x4  - st (goes in x21 and x20)
-  ;; x5  - offset (goes in x22)
+  //// x0 - stack pointer
+  //// x1 - unused
+  //// x2 - unused
+  //// x3 - function pointer
+  //// x4  - st (goes in x21 and x20)
+  //// x5  - offset (goes in x22)
 
-  ;; saved registers we need to modify for GOAL should be preserved
-  ; ARM64 requires 16-byte stack pointer alignment
+  //// saved registers we need to modify for GOAL should be preserved
+  // ARM64 requires 16-byte stack pointer alignment
   stp x20, x21, [sp, #-16]!
-  ;; also stash the current stack pointer on the stack
-  ;; NOTE - you cannot directly store or load the `sp` register in arm64
+  //// also stash the current stack pointer on the stack
+  //// NOTE - you cannot directly store or load the `sp` register in arm64
   mov x9, sp
   stp x22, x9, [sp, #-16]!
 
-  ;; switch to new stack
+  //// switch to new stack
   mov sp, x0
 
-  mov x20, x4 ;; set GOAL function pointer  
-  mov x21, x4 ;; symbol table
-  mov x22, x5 ;; offset
-  ;; call GOAL by function pointer
+  mov x20, x4 //// set GOAL function pointer  
+  mov x21, x4 //// symbol table
+  mov x22, x5 //// offset
+  //// call GOAL by function pointer
   blr x3
 
-  ;; restore registers
+  //// restore registers
   ldp x22, x9, [sp], #16
   mov sp, x9
   ldp x20, x21, [sp], #16
--- a/game/kernel/common/kscheme.cpp
+++ b/game/kernel/common/kscheme.cpp
@@ -84,7 +84,15 @@
 
 extern "C" {
 // defined in asm_funcs.asm
-#ifdef __linux__
+#ifdef __aarch64__
+uint64_t _call_goal_asm_arm64(u64 a0, u64 a1, u64 a2, void* fptr, void* st_ptr, void* offset);
+uint64_t _call_goal_on_stack_asm_arm64(u64 rsp,
+                                       u64 u0,
+                                       u64 u1,
+                                       void* fptr,
+                                       void* st_ptr,
+                                       void* offset);
+#elif defined(__linux__)
 uint64_t _call_goal_asm_systemv(u64 a0, u64 a1, u64 a2, void* fptr, void* st_ptr, void* offset);
 uint64_t _call_goal_on_stack_asm_systemv(u64 rsp,
                                          u64 u0,
@@ -116,7 +124,9 @@
   void* st_ptr = (void*)st;
 
   void* fptr = f.c();
-#ifdef __linux__
+#ifdef __aarch64__
+  return _call_goal_asm_arm64(a, b, c, fptr, st_ptr, offset);
+#elif defined(__linux__)
   return _call_goal_asm_systemv(a, b, c, fptr, st_ptr, offset);
 #elif defined __APPLE__ && defined __x86_64__
   return _call_goal_asm_systemv(a, b, c, fptr, st_ptr, offset);
@@ -132,7 +142,9 @@
   void* st_ptr = (void*)st;
 
   void* fptr = f.c();
-#ifdef __linux__
+#ifdef __aarch64__
+  return _call_goal_on_stack_asm_arm64(rsp, 0, 0, fptr, st_ptr, offset);
+#elif defined(__linux__)
   return _call_goal_on_stack_asm_systemv(rsp, 0, 0, fptr, st_ptr, offset);
 #elif defined __APPLE__ && defined __x86_64__
   return _call_goal_on_stack_asm_systemv(rsp, 0, 0, fptr, st_ptr, offset);
--- a/game/mips2c/mips2c_private.h
+++ b/game/mips2c/mips2c_private.h
@@ -20,7 +20,9 @@
 extern u8* g_ee_main_mem;
 
 extern "C" {
-#ifdef __linux__
+#ifdef __aarch64__
+u64 _call_goal8_asm_arm64(void* func, u64* arg_array, u64 zero, u64 pp, u64 st, void* off);
+#elif defined(__linux__)
 u64 _call_goal8_asm_systemv(void* func, u64* arg_array, u64 zero, u64 pp, u64 st, void* off);
 #elif defined __APPLE__ && defined __x86_64__
 u64 _call_goal8_asm_systemv(void* func, u64* arg_array, u64 zero, u64 pp, u64 st, void* off) asm(
@@ -353,7 +355,10 @@
     u64 args[8] = {gprs[a0].du64[0], gprs[a1].du64[0], gprs[a2].du64[0], gprs[a3].du64[0],
                    gprs[t0].du64[0], gprs[t1].du64[0], gprs[t2].du64[0], gprs[t3].du64[0]};
     ASSERT(addr);
-#ifdef __linux__
+#ifdef __aarch64__
+    gprs[v0].du64[0] = _call_goal8_asm_arm64(g_ee_main_mem + addr, args, 0, gprs[s6].du64[0],
+                                             gprs[s7].du64[0], g_ee_main_mem);
+#elif defined(__linux__)
     gprs[v0].du64[0] = _call_goal8_asm_systemv(g_ee_main_mem + addr, args, 0, gprs[s6].du64[0],
                                                gprs[s7].du64[0], g_ee_main_mem);
 #elif defined __APPLE__ && defined __x86_64__
--- a/game/mips2c/mips2c_table.cpp
+++ b/game/mips2c/mips2c_table.cpp
@@ -11,7 +11,9 @@
 #include "game/runtime.h"
 
 extern "C" {
-#ifdef __linux__
+#ifdef __aarch64__
+void _mips2c_call_arm64();
+#elif defined(__linux__)
 void _mips2c_call_systemv();
 #elif defined __APPLE__ && defined __x86_64__
 void _mips2c_call_systemv() asm("_mips2c_call_systemv");
@@ -719,7 +721,9 @@
     ptr++;
 
     // call the other function
-#ifdef __linux__
+#ifdef __aarch64__
+    addr = (u64)_mips2c_call_arm64;
+#elif defined(__linux__)
     addr = (u64)_mips2c_call_systemv;
 #elif defined __APPLE__ && defined __x86_64__
     addr = (u64)_mips2c_call_systemv;
--- a/game/graphics/opengl_renderer/SkyBlendCPU.cpp
+++ b/game/graphics/opengl_renderer/SkyBlendCPU.cpp
@@ -22,7 +22,7 @@
 }
 
 void blend_sky_initial_fast(u8 intensity, u8* out, const u8* in, u32 size) {
-#ifndef __arm64__
+#if !defined(__arm64__) && !defined(__aarch64__)
   if (get_cpu_info().has_avx2) {
 #ifdef __AVX2__
     __m256i intensity_vec = _mm256_set1_epi16(intensity);
@@ -53,7 +53,7 @@
 }
 
 void blend_sky_fast(u8 intensity, u8* out, const u8* in, u32 size) {
-#ifndef __arm64__
+#if !defined(__arm64__) && !defined(__aarch64__)
   if (get_cpu_info().has_avx2) {
 #ifdef __AVX2__
     __m256i intensity_vec = _mm256_set1_epi16(intensity);
@@ -205,4 +205,4 @@
     m_textures[i].tex = tex_pool.give_texture_and_load_to_vram(in, tbp);
     m_textures[i].tbp = tbp;
   }
-}
\ No newline at end of file
+}
--- a/game/runtime.cpp
+++ b/game/runtime.cpp
@@ -92,6 +92,12 @@
 BackgroundWorker g_background_worker;
 int g_server_port = DECI2_PORT;
 
+#ifdef MAP_32BIT
+constexpr int LOW_32BIT_MMAP_FLAG = MAP_32BIT;
+#else
+constexpr int LOW_32BIT_MMAP_FLAG = 0;
+#endif
+
 namespace {
 
 int g_argc = 0;
@@ -161,9 +167,9 @@
         (u8*)mmap((void*)0x10000000, EE_MAIN_MEM_SIZE, PROT_EXEC | PROT_READ | PROT_WRITE,
 #ifdef __APPLE__
                   // has no map_populate
-                  MAP_ANONYMOUS | MAP_32BIT | MAP_PRIVATE, 0, 0);
+                  MAP_ANONYMOUS | LOW_32BIT_MMAP_FLAG | MAP_PRIVATE, 0, 0);
 #else
-                  MAP_ANONYMOUS | MAP_32BIT | MAP_PRIVATE | MAP_POPULATE, 0, 0);
+                  MAP_ANONYMOUS | LOW_32BIT_MMAP_FLAG | MAP_PRIVATE | MAP_POPULATE, 0, 0);
 #endif
   } else {
     g_ee_main_mem =
@@ -239,6 +245,7 @@
       break;
     case GameVersion::JakX:
       jakx::goal_main(g_argc, g_argv);
+      break;
     default:
       ASSERT_MSG(false, "Unsupported game version");
   }
